qiskit/circuit/library/generalized_gates/linear_function.py

```diff
@@ -0,0 +1,160 @@
+        in numpy array format.
+
+        A linear function can be synthesized into CX and SWAP gates using the Patel–Markov–Hayes
+        algorithm. For efficiency, the internal n x n matrix is stored in the format expected
+        by cnot_synth, which is the big-endian (and not the little-endian) bit-ordering convention.
+
+        Args:
+            linear (list[list] or ndarray[bool] or QuantumCircuit):
+                either an n x n matrix, describing the linear function,
+                or a quantum circuit composed of linear gates only
+                (currently supported gates are CX and SWAP).
+
+            validate_input: if True, performs more expensive input validation checks,
+                such as checking that a given n x n matrix is invertible.
+
+        Raises:
+            CircuitError: if the input is invalid:
+                either a matrix is non {square, invertible},
+                or a quantum circuit contains non-linear gates.
+
+        **Example:** the circuit
+
+        .. parsed-literal::
+
+            q_0: ──■──
+                 ┌─┴─┐
+            q_1: ┤ X ├
+                 └───┘
+            q_2: ─────
+
+        is represented by a 3x3 linear matrix
+
+        .. math::
+
+                \begin{pmatrix}
+                    1 & 0 & 0 \\
+                    1 & 1 & 0 \\
+                    0 & 0 & 1
+                \end{pmatrix}
+
+        """
+        if not isinstance(linear, (list, np.ndarray, QuantumCircuit)):
+            raise CircuitError(
+                "A linear function must be represented either by a list, "
+                "a numpy array, or a quantum circuit with linear gates."
+            )
+
+        if isinstance(linear, QuantumCircuit):
+            # The following function will raise a CircuitError if there are nonlinear gates.
+            self._linear = _linear_quantum_circuit_to_mat(linear)
+
+        else:
+            # Normalize to numpy array (coercing entries to 0s and 1s)
+            ok = True
+            try:
+                self._linear = np.array(linear, dtype=bool, copy=True)
+            except ValueError:
+                ok = False
+
+            if not ok:
+                raise CircuitError("A linear function must be represented by a square matrix.")
+
+            # Check that the matrix is square
+            if not len(self._linear.shape) == 2 or self._linear.shape[0] != self._linear.shape[1]:
+                raise CircuitError("A linear function must be represented by a square matrix.")
+
+            # Optionally, check that the matrix is invertible
+            if validate_input:
+                det = np.linalg.det(linear) % 2
+                if not np.allclose(det, 1):
+                    raise CircuitError(
+                        "A linear function must be represented by an invertible matrix."
+                    )
+
+        super().__init__(name="linear_function", num_qubits=len(self._linear), params=[])
```

Comment:
Is there a reason to not store `linear` in `params` rather than in `self._linear`?

Reply:
No good reason whatsoever. Is it preferable that all gate-specific information is stored in `params`?